platform/mellanox/mlnx-platform-api/sonic_platform/modules_mgmt.py

```diff
@@ -469,55 +514,88 @@ def check_module_type(self, port, module_sm_obj, dynamic=False):
+            powercap = int.from_bytes(powercap_ba, "big")
+            return powercap
+        elif isinstance(xcvr_api, sff8636.Sff8636Api) or isinstance(xcvr_api, sff8436.Sff8436Api):
+            field = xcvr_api.xcvr_eeprom.mem_map.get_field(consts.POWER_CLASS_FIELD)
+            power_class_ba = xcvr_api.xcvr_eeprom.reader(field.get_offset(), field.get_size())
+            power_class_bits = {bit_id: int((power_class_ba[0] >> bit_id) & 0b1) for bit_id in [7, 6, 5, 1, 0]}
+            if (power_class_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (0, 0, 0, 0):
+                powercap = POWER_CLASS_1_MAX_POWER
+            elif (power_clשass_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (0, 1, 0, 0):
+                powercap = POWER_CLASS_2_MAX_POWER
+            elif (power_class_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (1, 0, 0, 0):
+                powercap = POWER_CLASS_3_MAX_POWER
+            elif (power_class_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (1, 1, 0, 0):
+                powercap = POWER_CLASS_4_MAX_POWER
+            elif (power_class_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (1, 1, 0, 1):
+                powercap = POWER_CLASS_5_MAX_POWER
+            elif (power_class_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (1, 1, 1, 0):
+                powercap = POWER_CLASS_6_MAX_POWER
+            elif (power_class_bits[7], power_class_bits[6], power_class_bits[1], power_class_bits[0]) == (1, 1, 1, 1):
+                powercap = POWER_CLASS_7_MAX_POWER
+            else:
+                logger.log_error("Invalid value for power class field: {}".format(power_class_ba))
+                module_sm_obj.set_final_state(STATE_ERROR_HANDLER)
+                return STATE_ERROR_HANDLER
+
+            if power_class_bits[5] == 1:
+                read_power_class_8_byte = xcvr_api.xcvr_eeprom.read_raw(107, 1)
+                powercap = max(read_power_class_8_byte, powercap)
+            return powercap
 
     def check_power_cap(self, port, module_sm_obj, dynamic=False):
         logger.log_info("enter check_power_cap port {} module_sm_obj {}".format(port, module_sm_obj))
         sfp = sfp_module.SFP(port)
         xcvr_api = sfp.get_xcvr_api()
-        field = xcvr_api.xcvr_eeprom.mem_map.get_field(consts.MAX_POWER_FIELD)
-        powercap_ba = xcvr_api.xcvr_eeprom.reader(field.get_offset(), field.get_size())
-        logger.log_info("check_power_cap got powercap bytearray {} for port {} module_sm_obj {}".format(powercap_ba, port, module_sm_obj))
-        powercap = int.from_bytes(powercap_ba, "big")
+        powercap = self.get_module_max_power(port, xcvr_api, module_sm_obj)
+        if powercap is STATE_ERROR_HANDLER:
```

Comment:
Not sure I understand this line. Could you please elaborate?

Reply:
STATE_ERROR_HANDLER can be returned from get_module_max_power() in case we deal with an SFF module and its EEPROM bits for the power class field do not match any known power class. I assume the EEPROM will always have valid values so I dont think there is a real chance to return STATE_ERROR_HANDLER from get_module_max_power(), but I didn't want to leave get_module_max_power() without any default value returned from it.